bosswave.cpp

```diff
@@ -225,29 +452,980 @@ void BW::subscribe(QString uri, Res<PMessage> on_msg, Res<QString> on_done)
+        {
+            QString v = f->getHeaderS("value");
+            on_done("", v);
+        }
+    });
+}
+
+void BW::resolveLongAlias(QString al, Res<QString, QByteArray, bool> on_done)
+{
+    auto f = agent()->newFrame(Frame::RESOLVE_ALIAS);
+    f->addHeader("longkey", al);
+
+    agent()->transact(this, f, [=](PFrame f, bool)
+    {
+        if (f->checkResponse(on_done, QByteArray(), false))
+        {
+            QByteArray v = f->getHeaderB("value");
+            on_done("", v, v.count('\0') == v.size());
+        }
+    });
+}
+
+void BW::resolveShortAlias(QString al, Res<QString, QByteArray, bool> on_done)
+{
+    auto f = agent()->newFrame(Frame::RESOLVE_ALIAS);
+    f->addHeader("shortkey", al);
+
+    agent()->transact(this, f, [=](PFrame f, bool)
+    {
+        if (f->checkResponse(on_done, QByteArray(), false))
+        {
+            QByteArray v = f->getHeaderB("value");
+            on_done("", v, v.count('\0') == v.size());
+        }
+    });
+}
+
+void BW::resolveEmbeddedAlias(QString al, Res<QString, QString> on_done)
+{
+    auto f = agent()->newFrame(Frame::RESOLVE_ALIAS);
+    f->addHeader("longkey", al);
+
+    agent()->transact(this, f, [=](PFrame f, bool)
+    {
+        if (f->checkResponse(on_done, QStringLiteral("")))
+        {
+            QString v = f->getHeaderS("value");
+            on_done("", v);
+        }
+    });
+}
+
+void BW::resolveRegistry(QString key, Res<QString, RoutingObject*, RegistryValidity> on_done)
+{
+    auto f = agent()->newFrame(Frame::RESOLVE_REGISTRY);
+    f->addHeader("key", key);
+
+    agent()->transact(this, f, [=](PFrame f, bool)
+    {
+        if (f->checkResponse(on_done, (RoutingObject*) nullptr, RegistryValidity::StateError))
+        {
+            QList<RoutingObject*> ros = f->getRoutingObjects();
+            if (ros.length() == 0)
+            {
+                on_done("", nullptr, RegistryValidity::StateError);
+                return;
+            }
+            QString valid = f->getHeaderS("validity");
+            RegistryValidity validity;
+
+            if (valid == "valid")
+            {
+                validity = RegistryValidity::StateValid;
+            }
+            else if (valid == "expired")
+            {
+                validity = RegistryValidity::StateExpired;
+            }
+            else if (valid == "revoked")
+            {
+                validity = RegistryValidity::StateRevoked;
+            }
+            else if (valid == "unknown")
+            {
+                validity = RegistryValidity::StateUnknown;
+            }
+            else
+            {
+                std::string rawvalid = valid.toStdString();
+                throw BadRouterMessageException("Invalid \"validity\" value", rawvalid.data());
+            }
+
+            on_done("", ros.first(), validity);
+        }
+    });
+}
+
+void BW::entityBalances(Res<QString, QVector<struct balanceinfo>> on_done)
```

Comment:
Is there a reason you chose qvector over qlist? I generally prefer keeping everything as qlist